File: 2016/Day19/main.dart
int part1(int numElves) {
  int highestPowerOf2 = 1 << (numElves.bitLength - 1);
  int l = numElves - highestPowerOf2;
  return 2 * l + 1;
}

int part2(int numElves) {
  int p3 = 1;
  while (p3 * 3 <= numElves) {
    p3 *= 3;
  }

  if (numElves == p3)
    return numElves;
  else if (numElves <= 2 * p3)
    return numElves - p3;
  else
    return 2 * numElves - 3 * p3;
}

void main() {
  const numElves = 3017957;

  print('Part 1: ${part1(numElves)}');
  print('Part 2: ${part2(numElves)}');
}

File: 2016/Day19/main.py
def part1(num_elves: int) -> int:
    highest_power_of_2 = 1 << (num_elves.bit_length() - 1)
    l = num_elves - highest_power_of_2
    return 2 * l + 1

def part2(num_elves: int) -> int:
    p3 = 1
    while p3 * 3 <= num_elves:
        p3 *= 3
    
    if num_elves == p3:
        return num_elves
    elif num_elves <= 2 * p3:
        return num_elves - p3
    else:
        return 2 * num_elves - 3 * p3

if __name__ == '__main__':
    num_elves = 3017957

    print(f'Part 1: {part1(num_elves)}')
    print(f'Part 2: {part2(num_elves)}')